File: verificaTipos.pl
%%Verificador de tipos em prolog

verificarTipos(_, valor_inteiro(X),tInt) :- integer(X).
verificarTipos(_, valor_booleano(true), tBool) :- !.
verificarTipos(_, valor_booleano(false), tBool) :- !.
%%TODO
%verificarTipos(Gamma, var(_), Res) :-

verificarTipos(_, valor_booleano(_), tErro).

verificarTipos(Gamma, soma(Lhs, Rhs), Res) :-
    verificarTipos(Gamma, Lhs, tInt),!,
    verificarTipos(Gamma, Rhs, tInt),!,
    Res = tInt.
verificarTipos(_, soma(_,_), tErro).

verificarTipos(Gamma, subtracao(Lhs, Rhs), Res) :-
    verificarTipos(Gamma, Lhs, tInt),!,
    verificarTipos(Gamma, Rhs, tInt),!,
    Res = tInt.
verificarTipos(_, subtracao(_,_), tErro).

verificarTipos(Gamma, multiplicacao(Lhs, Rhs), Res) :-
    verificarTipos(Gamma, Lhs, tInt),!,
    verificarTipos(Gamma, Rhs, tInt),!,
    Res = tInt.

verificarTipos(_,multiplicacao(_,_), tErro).

verificarTipos(_, divisao(_, valor_inteiro(0)), tErro):-!.

verificarTipos(Gamma, divisao(Lhs, Rhs), Res) :-
    verificarTipos(Gamma, Lhs, tInt),!,
    verificarTipos(Gamma, Rhs, tInt),!,
    Res = tInt.

verificarTipos(_, divisao(_,_), tErro).

verificarTipos(Gamma, lambda(Arg,Exp),tFuncao(Targ,Texp)) :-
    pesquisarAmbiente(Arg, Gamma, RArg),!,
    verificarTipos(Gamma,RArg, Targ),
    verificarTipos(Gamma,Exp, Texp).

verificarTipos(_,lambda(_,_), tErro).

verificarTipos(Gamma, var(X), Res) :-
    pesquisarAmbiente(X, Gamma, Exp),
    verificarTipos(Gamma, Exp, Res),!.

verificarTipos(Gamma, var(X), variavel_nao_encontrada) :-
    pesquisarAmbiente(X, Gamma, variavel_nao_encontrada),!.

verificarTipos(Gamma, let(X, Exp1, Exp2), T2) :-
    verificarTipos([(X,Exp1)|Gamma], Exp2, T2).

verificarTipos(Gamma,aplicacao(lambda(Arg,Exp),Exp2),tFuncao(Targ,Texp)):- !,
    verificarTipos([(Arg,Exp2)|Gamma], Exp, Texp),
    verificarTipos([(Arg,Exp2)|Gamma], Exp2, Targ).

verificarTipos(Gamma, aplicacao(Exp1,Exp2), Res):-
    pesquisarAmbiente(Exp1, Gamma, lambda(Arg,Exp)),!,
    verificarTipos([(Arg,Exp2)|Gamma], Exp, Res).

verificarTipos(_, aplicacao(_,_), tErro).

pesquisarAmbiente(_, [], variavel_nao_encontrada).
pesquisarAmbiente(Var, [(Var,Exp)|_], Exp) :- !.
pesquisarAmbiente(Var, [_|Tail], Res) :- pesquisarAmbiente(Var, Tail, Res).

File: VerificaTipos.hs
module VerificaTipos where 


-- Os tipos validos sao definidos com o tipo algebrico,
-- que pode ser um tipo inteiro, um tipo booleano e um tipo funcao. 
-- O tipo funcao deve expressar tanto o tipo do argumento
-- quanto o tipo do retorno. 
-- As expressoes, conforme mencionado anteriormente,
-- envolvem tanto valores inteiros quanto booleanos,
-- bem como expressoes binarias (soma, subtracao, etc.),
-- expressoes let, lambda, aplicacao de funcoes e {if-then-else}.


type Id = String

type Gamma= [(Id, Tipo)]

data Tipo = TInt
          | TBool
          | TFuncao Tipo Tipo 
    deriving(Show, Eq)

data Expressao = ValorI Int
               | ValorB Bool
               | Soma Expressao Expressao
               | Subtracao Expressao Expressao 
               | Multiplicacao Expressao Expressao
               | Divisao Expressao Expressao 
               | Let Id Expressao Expressao       
               | Ref Id
               | Lambda (Id, Tipo) Tipo Expressao
               | Aplicacao Expressao Expressao
               | If Expressao Expressao Expressao
    deriving(Show, Eq)


verificarTipos :: Expressao -> Gamma -> Maybe Tipo
verificarTipos (ValorI n)                   _ = return TInt
verificarTipos (ValorB b)                   _ = return TBool
verificarTipos (Lambda (v, t1) t2 exp)  gamma = return (TFuncao t1 t2)
verificarTipos (Soma l r)               gamma = expAlgebricas l r gamma
verificarTipos (Subtracao l r)          gamma = expAlgebricas l r gamma
verificarTipos (Multiplicacao l r)      gamma = expAlgebricas l r gamma
--verificarTipos (Divisao l r)            gamma = expAlgebricas l r gamma
verificarTipos (Divisao l r)            gamma = -- retorna Nothing se o denominador da divisão for um valor zero
    if r == (ValorI 0)
        then Nothing
        else expAlgebricas l r gamma
verificarTipos (Ref v)                  gamma = pesquisar v gamma

verificarTipos (Let v e c)              gamma =
    verificarTipos e gamma  >>= \lt ->
    verificarTipos c gamma2 >>= \rt ->
    if lt == rt
        then return rt
        else Nothing
    where
        gamma2 = incrementaAmb v (verificarTipos e gamma) gamma

--verificarTipos (Aplicacao def arg)      gamma = undefined
verificarTipos (Aplicacao def arg)      gamma =
--    let gamma' = [(v, tId)]
    case def of
        (Lambda (v, tId) tExp exp) -> verificarTipos arg gamma >>= \a ->
            if a == tId
                then verificarTipos exp gamma'
                else Nothing
            where
                gamma' = [(v, tId)]
        otherwise -> error ("Aplicacao de funcao nao anonima")

{-
verificarTipos (Aplicacao def arg)      gamma =
    verificarTipos def gamma >>= \d ->
    case d of
        (TFuncao tId tExp) -> verificarTipos arg gamma >>= \a ->
            if a == tId
                then verificarTipos tExp -- completar
                else Nothing
        otherwise -> error ("Aplicacao de funcao nao anonima")
-}
--verificarTipos (If test pass fail)      gamma = undefined
verificarTipos (If test pass fail)      gamma =
    verificarTipos test gamma >>= \t ->
    if t == TBool
        then verificarTipos pass gamma >>= \p ->
             verificarTipos fail gamma >>= \f ->
             if p == f
                then return p
                else Nothing
        else Nothing


-- Verifica se os tipos equivalem a TInt.
expAlgebricas :: Expressao -> Expressao -> Gamma -> Maybe Tipo
expAlgebricas l r gamma =
    verificarTipos l gamma >>= \lt ->
    verificarTipos r gamma >>= \rt ->
    if lt == TInt && rt == TInt
        then return TInt
        else Nothing


-- Realiza uma pesquisa por uma determinada variaval em um ambiente.
pesquisar :: Id -> Gamma -> Maybe Tipo
pesquisar v [] = error "Variavel nao declarada."
pesquisar v ((i,e):xs)
    | v == i = return e
    | otherwise = pesquisar v xs


-- Incrementa o ambiente.
incrementaAmb :: Id -> Maybe Tipo -> Gamma -> Gamma
incrementaAmb n Nothing  [] = []
incrementaAmb n Nothing  ((i,e):xs) = ((i,e):xs)
incrementaAmb n (Just v) [] = [(n, v)]
incrementaAmb n (Just v) ((i,e):xs)
    | n == i = incrementaAmb n (return v) []
    | otherwise = incrementaAmb n (return v) xs
